core/src/main/java/org/elasticsearch/index/cache/IndexCacheModule.java

```diff
@@ -35,24 +35,26 @@
     // for test purposes only
     public static final String QUERY_CACHE_EVERYTHING = "index.queries.cache.everything";
 
-    private final Settings settings;
+    private Class<? extends QueryCache> queryCacheClass;
 
     public IndexCacheModule(Settings settings) {
-        this.settings = settings;
-    }
-
-    @Override
-    protected void configure() {
         String queryCacheType = settings.get(QUERY_CACHE_TYPE, INDEX_QUERY_CACHE);
-        Class<? extends QueryCache> queryCacheImpl;
         if (queryCacheType.equals(INDEX_QUERY_CACHE)) {
-            queryCacheImpl = IndexQueryCache.class;
+            queryCacheClass = IndexQueryCache.class;
         } else if (queryCacheType.equals(NONE_QUERY_CACHE)) {
-            queryCacheImpl = NoneQueryCache.class;
+            queryCacheClass = NoneQueryCache.class;
         } else {
             throw new IllegalArgumentException("Unknown QueryCache type [" + queryCacheType + "]");
         }
-        bind(QueryCache.class).to(queryCacheImpl).in(Scopes.SINGLETON);
+    }
+
+    public void setQueryCacheClass(Class<? extends QueryCache> queryCacheClass) {
```

Comment:
It still can. The plugin just needs to enforce that the user settings cannot contain that setting (so that the additional settings are not overriden by the user settings). This can be done by throwing an exception on init.